### PR TITLE
Incremental Export and User by Role support

### PR DESCRIPTION
**Pull request by Hillgod**
Added support for the incremental export API.
Also added support for querying for users by role.

Updated README.md to reflect the incremental export API.

### CODE REVIEW

**Comment by johnou** [comment]
Rate limiting is an interesting one, I'd like to hear @stephenc 's thoughts on it but I am happy to share our rate limiting solution.

```java
AsyncHttpClientConfig config = new AsyncHttpClientConfig.Builder().addRequestFilter(new RateLimiterRequestFilter()).build();
```

```java
/**
 * {@link com.ning.http.client.filter.RequestFilter} which is tuned to Zendesk's API rate limit (200 rpm).
 *
 * @author Johno Crawford (johno@sulake.com)
 */
public class SimpleRateLimiterRequestFilter implements RequestFilter {

    private RateLimiter rateLimiter = RateLimiter.create(3.0);

    @Override
    public FilterContext filter(FilterContext context) throws FilterException {
        rateLimiter.acquire();
        return new FilterContext.FilterContextBuilder(context).build();
    }
}
```

```java
/**
 * {@link com.ning.http.client.filter.RequestFilter} which is tuned to Zendesk's API rate limit (200 rpm / endpoint).
 *
 * @author Johno Crawford (johno@sulake.com)
 */
public class RateLimiterRequestFilter implements RequestFilter {

    private static final Logger logger = Logger.getLogger(RateLimiterRequestFilter.class);

    private static final ConcurrentHashMap<byte[], RateLimiter> rateLimiters = new ConcurrentHashMap<>();

    @Override
    public FilterContext filter(FilterContext context) throws FilterException {
        RateLimiter rateLimiter = getRateLimiter(context.getRequest().getInetAddress().getAddress());
        double secondsWaited = rateLimiter.acquire();
        if (secondsWaited > 0) {
            logger.info("Request throttled, waited " + secondsWaited + " seconds.");
        }
        return new FilterContext.FilterContextBuilder(context).build();
    }

    private RateLimiter getRateLimiter(byte[] remoteAddress) {
        RateLimiter rateLimiter = rateLimiters.get(remoteAddress);
        if (rateLimiter != null) {
            return rateLimiter;
        }
        rateLimiter = RateLimiter.create(3.0);
        RateLimiter oldRateLimiter = rateLimiters.putIfAbsent(remoteAddress, rateLimiter);
        return oldRateLimiter != null ? oldRateLimiter : rateLimiter;
    }
}
```

**Comment by Hillgod** [comment]
Egh, I didn't intend for that to be part of this pull request. It's a fairly naive implementation of it. I'll see what I can do downstream to get it out of this pull request, at least.

**Comment by stephenc** [comment]
@johnou your rate limiter looks good

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 1]
```diff
@@ -1195,9 +1240,11 @@ public T onCompleted(Response response) throws Exception {
     }
 
     private static final String NEXT_PAGE = "next_page";
+    private static final String END_TIME = "end_time";
+    private static final long FIVE_MINUTES_IN_MS = 300000;
```
Comment (johnou) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: Might be a little nicer to read using TimeUnit eg. TimeUnit.MINUTES.toMillis(5)

Comment (Hillgod) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: Cool, made that change.

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 2]
```diff
@@ -434,6 +462,23 @@ public void deleteTrigger(long triggerId) {
...
         return new PagedIterable<User>(cnst("/users.json"), handleList(User.class, "users"));
     }
+    
+    public Iterable<User> getUsersByRole(String role, String... roles) {
+        // Going to have to build this URI manually, because the RFC6570 template spec doesn't support
+        // variables like ?role[]=...role[]=..., which is what Zendesk requires.
+        // See https://developer.zendesk.com/rest_api/docs/core/users#filters
+        final StringBuilder uriBuilder = new StringBuilder("/users.json");
+        if (roles.length == 0) {
+            uriBuilder.append("?role=").append(role);
```
Comment (johnou) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: do we need to uri encode the role?

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 3]
```diff
@@ -228,6 +242,20 @@ public void deleteTickets(long id, long... ids) {
...
+    public Iterable<Ticket> getTicketsIncrementally(Date startTime) {
+        return new PagedIterable<Ticket>(
+                tmpl("/incremental/tickets.json{?start_time}").set("start_time", (int) (startTime.getTime() / 1000)), 
+                handleIncrementalList(Ticket.class, "tickets"));                
+    }
+    
+    public Iterable<Ticket> getTicketsIncrementally(Date startTime, Date endTime) {
+        return new PagedIterable<Ticket>(
+                tmpl("/incremental/tickets.json{?start_time,end_time}")
+                    .set("start_time", (int) (startTime.getTime() / 1000))
```
Comment (johnou) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: for readability I'd suggest creating a small private method which maybe takes a date divides by 1000 and casts, looks a ugly over and over.

**Comment by johnou** [comment]
When I get some time I'll look at trying to integrate something like http://blog.nirav.name/2015/02/a-simple-rate-limiter-using-javas.html into the project instead of adding a dependency on guava..

**Comment by johnou** [comment]
@Hillgod would you have time to implement the remaining endpoints listed https://developer.zendesk.com/rest_api/docs/core/incremental_export if not please mark which ones are not implemented in the README.

**Comment by stephenc** [comment]
@johnou yes, let's not add guava as a dependency... that project has a completely different idea of backwards compatibility from everyone else... depending on it in a utility jar (such as this) is a really bad plan... such a pity that we have to run NIH syndrome with guava as it would otherwise be a lovely library... sadly I have been burned 5 times too often with guava

**Comment by Hillgod** [comment]
I'll make the suggested code changes ASAP. I'm about to be on vacation for over a week, so the other incremental exports will have to wait until at least then, but I'm probably up for it. Regardless, I'll update the README.md to better reflect the new situation w.r.t that API.

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 4]
```diff
@@ -434,6 +462,23 @@ public void deleteTrigger(long triggerId) {
...
+        final StringBuilder uriBuilder = new StringBuilder("/users.json");
+        if (roles.length == 0) {
+            uriBuilder.append("?role=").append(role);
+        } else {
+            uriBuilder.append("?role[]=").append(role);
+        }
+        for (final String curRole : roles) {
+            uriBuilder.append("&role[]=").append(curRole);
+        }
+        logger.info(uriBuilder.toString());
```
Comment (johnou) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: remove debug logging?

**Comment by Hillgod** [comment]
I think I've addressed all concerns. Please let me know if you'd like to be do anything else before merging this in. Thanks!

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 5]
```diff
@@ -1212,23 +1260,53 @@ public String getNextPage() {
...
+                if (node == null) {
+                    throw new NullPointerException(NEXT_PAGE + " property not found, pagination not supported" +
+                            (clazz != null ? " for " + clazz.getName() : ""));
+                }
+                JsonNode endTimeNode = responseNode.get(END_TIME);
+                if (endTimeNode == null || endTimeNode.asLong() == 0) {
+                    throw new NullPointerException(END_TIME + " property not found, incremental export pagination not supported" +
+                            (clazz != null ? " for " + clazz.getName() : ""));
+                }
+                if (endTimeNode.asLong() * 1000 > System.currentTimeMillis() - FIVE_MINUTES_IN_MS) {
```
Comment (stephenc) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: You could use `TimeUnit.SECONDS.toMillis(endTimeNode.asLong())` rather than the less expressive `* 1000`

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 6]
```diff
@@ -1294,6 +1372,20 @@ private void logResponse(Response response) throws IOException {
...
+        try {
+            return URLEncoder.encode(str, UTF_8);
+        } catch (UnsupportedEncodingException e) {
+            // This really can't happen. UTF-8 is guaranteed to exist.
+            return str;
+        }
+    }
+
+    private static int msToSeconds(long millis) {
+        return (int) (millis / 1000);
```
Comment (stephenc) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: I'd rather `TimeUnit.MILLISECONDS.toSeconds()` which can safely be in-lined

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 7]
```diff
@@ -1528,8 +1620,8 @@ public Builder setOauthToken(String oauthToken) {
             }
             return this;
         }
-
-
+
```
Comment (stephenc) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: unnecessary whitespace changes

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 8]
```diff
@@ -1195,9 +1241,11 @@ public T onCompleted(Response response) throws Exception {
     }
 
     private static final String NEXT_PAGE = "next_page";
+    private static final String END_TIME = "end_time";
+    private static final long FIVE_MINUTES_IN_MS = TimeUnit.MINUTES.toMillis(5);
 
     private abstract class PagedAsyncCompletionHandler<T> extends AsyncCompletionHandler<T> {
-        private String nextPage;
+        String nextPage;
```
Comment (stephenc) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: does this really need escalation to `/*package*/` scope? if so, I'd prefer a getter/setter and keep it `private`

Comment (Hillgod) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: The reason I just made to `package` scope was we're already dealing with a private class, and an anonymous class below needs to set the value. But I've made it private and added a setter.

**Comment by johnou** [comment]
Merged, thanks @Hillgod.
